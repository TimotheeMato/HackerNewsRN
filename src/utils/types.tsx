import { Action } from 'redux';

export interface RootState {
  topstories: StoriesState;
  beststories: StoriesState;
  newstories: StoriesState;
}

export interface StoriesState {
  isFetching: boolean;
  error: object | null;
  stories: Story[] | null;
  ids: string[] | null;
}

export interface ActionPayload<T = any> extends Action {
  type: string;
  payload: T;
}

export interface ActionStory extends ActionPayload {
  category: StoryCategory;
}

export enum StoryCategory {
  TOP_STORIES = 'topstories',
  BEST_STORIES = 'beststories',
  NEW_STORIES = 'newstories',
}

export interface Story {
  id: number;
  by: string;
  kids: number[] | null;
  score: number;
  time: number;
  title: string;
  type: string;
  url: string;
}