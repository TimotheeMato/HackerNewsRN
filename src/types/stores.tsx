export interface IStoriesStateType {
  isFetching: boolean;
  error: object | null;
  stories: object | null;
}
