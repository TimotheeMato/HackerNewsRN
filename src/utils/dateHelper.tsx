export function timeSince(date: Date) {
  const seconds = Math.floor((new Date().valueOf() - date.valueOf()) / 1000);
  let interval = Math.floor(seconds / 31536000);

  if (interval > 1) {
    return interval + ' years';
  }
  interval = Math.floor(seconds / 2592000);
  if (interval > 1) {
    return interval + ' months';
  }
  interval = Math.floor(seconds / 86400);
  if (interval > 1) {
    return interval + ' days';
  } else if (interval === 1) {
    return interval + ' day';
  }
  interval = Math.floor(seconds / 3600);
  if (interval > 1) {
    return interval + ' hours';
  } else if (interval === 1) {
    return interval + ' hour';
  }
  interval = Math.floor(seconds / 60);
  if (interval > 1) {
    return interval + ' minutes';
  } else if (interval === 1) {
    return interval + ' minute';
  }
  return Math.floor(seconds) + ' seconds';
}