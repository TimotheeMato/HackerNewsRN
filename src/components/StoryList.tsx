import React from 'react';
import {
  View,
  Text,
  ActivityIndicator,
  FlatList,
  StyleSheet,
  TouchableOpacity,
  Linking,
  Share,
  Platform,
} from 'react-native';
import FontAwesome5 from 'react-native-vector-icons/FontAwesome5';
import { connect } from 'react-redux';

import { RootState, Story, StoryCategory } from '../utils/types';
import { fetchStories, fetchStoriesIds } from '../redux/actions/storiesActions';
import { timeSince } from '../utils/dateHelper';

interface OwnProps {
  storyCategory: StoryCategory;
  store?: any; // Necessary for test connected component
}

interface StateProps {
  isFetching: boolean;
  error: object | null;
  stories: Story[] | null;
  ids: string[] | null;
}

interface DispatchProps {
  fetchStoriesIds: () => void;
  fetchStories: (ids: string[]) => void;
}

export type Props = StateProps & DispatchProps & OwnProps;

interface State {
  index: number;
}

const PAGE_SIZE = 25;

export class StoryList extends React.Component<Props, State> {

  static goToUrl(itemUrl: string) {
    Linking.openURL(itemUrl);
  }

  static shareUrl(item: Story) {
    const itemUrl = item.url ? item.url : 'https://news.ycombinator.com/item?id=' + item.id;
    let message = item.title;
    message += Platform.OS === 'android' ? ` ${itemUrl}` : '';
    Share.share({
      message,
      url: itemUrl,
      title: 'Check out this link',
    }, {
      // Android only:
      dialogTitle: 'Check out this link',
    });
  }

  constructor(props: Props) {
    super(props);
    this.state = {
      index: 0,
    };
  }

  componentDidMount() {
    this.props.fetchStoriesIds();
  }

  componentDidUpdate(prevProps: Props) {
    if (!prevProps.ids && this.props.ids) {
      const ids = this.props.ids.slice(0, PAGE_SIZE);
      this.props.fetchStories(ids);
    }
    if (this.props.stories && prevProps.stories !== this.props.stories) {
      this.setState({ index: this.props.stories.length });
    }
  }

  loadMoreStories = () => {
    if (this.props.ids) {
      const ids = this.props.ids.slice(this.state.index, this.state.index + PAGE_SIZE);
      this.props.fetchStories(ids);
    }
  };

  keyExtractor = (item: any) => `list-item-${item.id}`;

  renderItem = (data: any) => {
    const url = require('url');
    let urlHostname;
    if (data.item.url) {
      urlHostname = url.parse(data.item.url).hostname;
      urlHostname = urlHostname.startsWith('www.') ? urlHostname.substring(4) : urlHostname;
    }
    const date = new Date(data.item.time * 1000);
    return (
      <View style={styles.listItem}>
          <Text style={styles.date}>{`${timeSince(date)} ago by ${data.item.by}`}</Text>
        <TouchableOpacity onPress={() => StoryList.goToUrl(data.item.url)}>
          <Text style={styles.title} testID={'item-title'}>{data.item.title}</Text>
          {urlHostname && <Text style={styles.url}>{`(${urlHostname})`}</Text>}
        </TouchableOpacity>
        <View style={styles.bottomRow}>
          <TouchableOpacity style={styles.bottomItem}>
            <FontAwesome5 name={'arrow-alt-circle-up'} size={18} style={styles.bottomIcon} brand/>
            <Text style={styles.bottomText}>{data.item.score}</Text>
          </TouchableOpacity>
          <TouchableOpacity style={styles.bottomItem}>
            <FontAwesome5 name={'comments'} size={16} style={styles.bottomIcon} brand/>
            <Text style={styles.bottomText}>{data.item.kids ? data.item.kids.length : '0'}</Text>
          </TouchableOpacity>
          <TouchableOpacity style={styles.bottomItem} onPress={() => StoryList.shareUrl(data.item)}>
            <FontAwesome5 name={'share-alt'} size={18} style={styles.bottomIcon} brand/>
            <Text style={styles.bottomText}>Share</Text>
          </TouchableOpacity>
        </View>
      </View>
    );
  };

  renderFooterList = () => {
    return (
      <View>
        <ActivityIndicator style={styles.activityIndicator} size="large"/>
      </View>
    );
  };

  renderList() {
    return (
      <FlatList
        data={this.props.stories}
        keyExtractor={this.keyExtractor}
        renderItem={this.renderItem}
        refreshing={this.props.isFetching}
        onEndReached={this.loadMoreStories}
        onEndReachedThreshold={1}
        ListFooterComponent={this.renderFooterList}
      />
    );
  }

  renderContent() {
    if (this.props.isFetching && !this.props.stories) {
      return <ActivityIndicator size="large" style={styles.activityIndicator}/>;
    } else if (this.props.stories) {
      return this.renderList();
    } else  {
      return <Text>Error</Text>;
    }
  }

  render() {
    return (
      <View>
        {this.renderContent()}
      </View>
    );
  }
}

const mapStateToProps = (state: RootState, ownProps: OwnProps): StateProps => {
  const storyState = state[ownProps.storyCategory];
  return {
    ...storyState,
  };
};

const mapDispatchToProps = (dispatch: any, ownProps: OwnProps): DispatchProps => ({
  fetchStoriesIds: () => dispatch(fetchStoriesIds(ownProps.storyCategory)),
  fetchStories: (ids: string[]) => dispatch(fetchStories(ids, ownProps.storyCategory)),
});

export default connect(mapStateToProps, mapDispatchToProps)(StoryList);

const styles = StyleSheet.create({
  listItem: {
    backgroundColor: 'white',
    marginTop: 10,
    marginHorizontal: 10,
    padding: 10,
    borderRadius: 3,
    borderWidth: 1,
    borderColor: '#E2E2E7',
  },
  activityIndicator: {
    padding: 10,
  },
  date: {
    fontSize: 11,
  },
  title: {
    fontWeight: '500',
    fontSize: 15,
    marginTop: 6,
  },
  url: {
    fontWeight: '500',
    fontSize: 14,
    marginTop: 6,
    color: '#b0afb3',
  },
  bottomRow: {
    flex: 1,
    flexDirection: 'row',
    marginTop: 14,
  },
  bottomItem: {
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'center',
  },
  bottomText: {
    color: '#b0afb3',
    fontWeight: '500',
    fontSize: 13,
  },
  bottomIcon: {
    marginRight: 6,
    color: '#b0afb3',
    fontWeight: '500',
  },
});
