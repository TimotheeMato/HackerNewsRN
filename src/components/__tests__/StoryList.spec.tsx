import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import { ActivityIndicator, FlatList, Text, Linking, Share } from 'react-native';

import { StoryList } from '../StoryList';
import { Story, StoryCategory } from '../../utils/types';

const stories: Story[] = [
  {
    id: 42,
    by: 'tmato',
    kids: [1, 2],
    score: 42,
    time: Math.round((new Date()).getTime() / 1000),
    title: 'This is a random title',
    type: 'story',
    url: 'http://google.com',
  },
  {
    id: 84,
    by: 'matoti',
    kids: [1, 2],
    score: 84,
    time: Math.round((new Date()).getTime() / 1000),
    title: 'This is a new random title',
    type: 'story',
    url: 'http://google.com/search',
  },
];
const ids: string[] = ['42', '84'];
const defaultProps = {
  isFetching: false,
  error: null,
  stories,
  ids,
  fetchStoriesIds: jest.fn(),
  fetchStories: jest.fn(),
  storyCategory: StoryCategory.TOP_STORIES,
};
Linking.openURL = jest.fn();
Share.share = jest.fn();

describe('StoryList', () => {
  let wrapper: ShallowWrapper;

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders correctly', () => {
    wrapper = shallow(
      <StoryList {...defaultProps} />,
    );
    expect(wrapper).toMatchSnapshot();
  });
  it('displays an error', () => {
    wrapper = shallow(
      <StoryList
        isFetching={false}
        error={{ message: 'error'}}
        stories={null}
        ids={null}
        fetchStoriesIds={jest.fn()}
        fetchStories={jest.fn()}
        storyCategory={StoryCategory.TOP_STORIES}
      />,
    );
    expect(wrapper.contains(<Text>Error</Text>)).toEqual(true);
  });
  it('loads stories', () => {
    wrapper = shallow(
      <StoryList
        isFetching={true}
        error={null}
        stories={null}
        ids={null}
        fetchStoriesIds={defaultProps.fetchStoriesIds}
        fetchStories={defaultProps.fetchStories}
        storyCategory={StoryCategory.TOP_STORIES}
      />,
    );
    expect(defaultProps.fetchStoriesIds).toHaveBeenCalled();
    wrapper.setProps({ ids });
    wrapper.update();
    expect(defaultProps.fetchStories).toHaveBeenCalled();
    expect(wrapper.find(ActivityIndicator)).toHaveLength(1);
    expect(wrapper.find(FlatList)).toHaveLength(0);
    wrapper.setProps({ stories });
    wrapper.update();
    expect(wrapper.find(FlatList)).toHaveLength(1);
  });
  it('loads more stories', () => {
    wrapper = shallow(
      <StoryList
        isFetching={true}
        error={null}
        stories={null}
        ids={null}
        fetchStoriesIds={defaultProps.fetchStoriesIds}
        fetchStories={defaultProps.fetchStories}
        storyCategory={StoryCategory.TOP_STORIES}
      />,
    );
    expect(defaultProps.fetchStoriesIds).toHaveBeenCalled();
    wrapper.setProps({ ids });
    wrapper.update();
    expect(defaultProps.fetchStories).toHaveBeenCalled();
    expect(wrapper.find(FlatList)).toHaveLength(0);
    wrapper.setProps({ stories });
    wrapper.update();
    expect(wrapper.find(FlatList)).toHaveLength(1);
  });
  it('renderItem', () => {
    wrapper = shallow(
      <StoryList {...defaultProps} />,
    );
    const component = wrapper.instance() as StoryList;
    expect(component.renderItem({ item: stories[0] })).toBeDefined();
  });
  it('loadsMoreStories', () => {
    wrapper = shallow(
      <StoryList {...defaultProps} />,
    );
    const component = wrapper.instance() as StoryList;
    component.loadMoreStories();
    expect(defaultProps.fetchStories).toHaveBeenCalled();
  });
  it('openUrl', () => {
    StoryList.goToUrl('google.com');
    expect(Linking.openURL).toHaveBeenCalled();
  });
  it('shareUrl', () => {
    StoryList.shareUrl(stories[0]);
    expect(Share.share).toHaveBeenCalled();
  });
});
