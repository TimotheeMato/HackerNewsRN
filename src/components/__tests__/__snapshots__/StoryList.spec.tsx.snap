// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StoryList renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <StoryList
    error={null}
    fetchStories={[MockFunction]}
    fetchStoriesIds={
      [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    ids={
      Array [
        "42",
        "84",
      ]
    }
    isFetching={false}
    navigation={
      Object {
        "navigate": [MockFunction],
      }
    }
    stories={
      Array [
        Object {
          "by": "tmato",
          "id": 42,
          "kids": Array [
            1,
            2,
          ],
          "score": 42,
          "time": 1539934494,
          "title": "This is a random title",
          "type": "story",
          "url": "http://google.com",
        },
        Object {
          "by": "matoti",
          "id": 84,
          "kids": Array [
            1,
            2,
          ],
          "score": 84,
          "time": 1539934494,
          "title": "This is a new random title",
          "type": "story",
          "url": "http://google.com/search",
        },
      ]
    }
    storyCategory="topstories"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": <FlatList
        ListFooterComponent={[Function]}
        data={
          Array [
            Object {
              "by": "tmato",
              "id": 42,
              "kids": Array [
                1,
                2,
              ],
              "score": 42,
              "time": 1539934494,
              "title": "This is a random title",
              "type": "story",
              "url": "http://google.com",
            },
            Object {
              "by": "matoti",
              "id": 84,
              "kids": Array [
                1,
                2,
              ],
              "score": 84,
              "time": 1539934494,
              "title": "This is a new random title",
              "type": "story",
              "url": "http://google.com/search",
            },
          ]
        }
        disableVirtualization={false}
        horizontal={false}
        initialNumToRender={10}
        keyExtractor={[Function]}
        maxToRenderPerBatch={10}
        numColumns={1}
        onEndReached={[Function]}
        onEndReachedThreshold={2}
        refreshing={false}
        renderItem={[Function]}
        scrollEventThrottle={50}
        updateCellsBatchingPeriod={50}
        windowSize={21}
      />,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "ListFooterComponent": [Function],
        "data": Array [
          Object {
            "by": "tmato",
            "id": 42,
            "kids": Array [
              1,
              2,
            ],
            "score": 42,
            "time": 1539934494,
            "title": "This is a random title",
            "type": "story",
            "url": "http://google.com",
          },
          Object {
            "by": "matoti",
            "id": 84,
            "kids": Array [
              1,
              2,
            ],
            "score": 84,
            "time": 1539934494,
            "title": "This is a new random title",
            "type": "story",
            "url": "http://google.com/search",
          },
        ],
        "disableVirtualization": false,
        "horizontal": false,
        "initialNumToRender": 10,
        "keyExtractor": [Function],
        "maxToRenderPerBatch": 10,
        "numColumns": 1,
        "onEndReached": [Function],
        "onEndReachedThreshold": 2,
        "refreshing": false,
        "renderItem": [Function],
        "scrollEventThrottle": 50,
        "updateCellsBatchingPeriod": 50,
        "windowSize": 21,
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <FlatList
          ListFooterComponent={[Function]}
          data={
            Array [
              Object {
                "by": "tmato",
                "id": 42,
                "kids": Array [
                  1,
                  2,
                ],
                "score": 42,
                "time": 1539934494,
                "title": "This is a random title",
                "type": "story",
                "url": "http://google.com",
              },
              Object {
                "by": "matoti",
                "id": 84,
                "kids": Array [
                  1,
                  2,
                ],
                "score": 84,
                "time": 1539934494,
                "title": "This is a new random title",
                "type": "story",
                "url": "http://google.com/search",
              },
            ]
          }
          disableVirtualization={false}
          horizontal={false}
          initialNumToRender={10}
          keyExtractor={[Function]}
          maxToRenderPerBatch={10}
          numColumns={1}
          onEndReached={[Function]}
          onEndReachedThreshold={2}
          refreshing={false}
          renderItem={[Function]}
          scrollEventThrottle={50}
          updateCellsBatchingPeriod={50}
          windowSize={21}
        />,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "ListFooterComponent": [Function],
          "data": Array [
            Object {
              "by": "tmato",
              "id": 42,
              "kids": Array [
                1,
                2,
              ],
              "score": 42,
              "time": 1539934494,
              "title": "This is a random title",
              "type": "story",
              "url": "http://google.com",
            },
            Object {
              "by": "matoti",
              "id": 84,
              "kids": Array [
                1,
                2,
              ],
              "score": 84,
              "time": 1539934494,
              "title": "This is a new random title",
              "type": "story",
              "url": "http://google.com/search",
            },
          ],
          "disableVirtualization": false,
          "horizontal": false,
          "initialNumToRender": 10,
          "keyExtractor": [Function],
          "maxToRenderPerBatch": 10,
          "numColumns": 1,
          "onEndReached": [Function],
          "onEndReachedThreshold": 2,
          "refreshing": false,
          "renderItem": [Function],
          "scrollEventThrottle": 50,
          "updateCellsBatchingPeriod": 50,
          "windowSize": 21,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
